engine/src/main/java/org/terasology/config/flexible/FlexibleConfig.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.config.flexible;
+
+import com.google.common.collect.Maps;
+import org.terasology.assets.ResourceUrn;
+
+import java.util.Map;
+
+public class FlexibleConfig {
+    private Map<Key<?>, Setting> settingMap;
+
+    public FlexibleConfig() {
+        this.settingMap = Maps.newHashMap();
+    }
+
+    public <V> Key<V> add(Setting<V> setting) {
+        Key<V> key = new Key<>(setting.getId(), setting.getValue());
+
+        // Maybe throw an exception?
+        if (has(key))
+            return null;
```

Comment:
The key is the SimpleUri stored in the setting, so we don't need the config to create it and return it.

Of course we need to make sure that the map is aware of the fact that

`new SimpleUri("module:setting") == new SimpleUri("module:setting")`

that is, SimpleUris are considered identical if their content is identical.